snowpackjs/snowpack
# upgrade svelte-hmr@0.12 (breaking, changed default), and update csa S…

## Changes

Upgrade `svelte-hmr` to 0.12 in `plugin-svelte`.

This is a **breaking change** for the plugin, because the names and defaults of some options have changed.

Namely, following [this discussion](https://github.com/snowpackjs/snowpack/discussions/1567), preservation of local state has been disabled by default. Similar confusion around preservation of whole local state have arisen multiple times already, so the behaviour has been deemed too confusing. It's better to have "just working" HMR.

Moreover, preservation of local state is not actually all that useful in real life scenarios. For the rare instances where you may want it, a magic `// @hmr:keep` comment has been added, that allows to preserve the state of variables individually, instead all the variables in the component. This is far easier to understand instinctively than the reverse.

I also hope people might notice that the counter kept running in demos / templates, and come back to them to find why when they need it. Whereas they won't know what to search for when they face the weird behaviours like the one in the discussion.

I have also updated the example Svelte component in the Svelte and Svelte TS create-snowpack-app templates to keep the counter running, and demonstrate the magic HMR comment. The indentation in the TS template was off for some reasons, so I fixed it.

## Testing

For both Svelte and Svelte TS:
- create an app with csa
- install my local `@snowpack/svelte-plugin`
- confirm that the counter is now reset on HMR updates
- add the magic `@hmr:keep` comment
- confirm that the counter's state is now preserved across updates

## Docs

The goal of this change is in big part to make Svelte HMR a zero-options / no-docs-needed feature. I hesitated to add a note about `// @hmr:keep`, but I guess demonstrating it with the counter in the template is enough, and probably better.
